### PR TITLE
Add support for Symfony 5 / admin 4

### LINKED ISSUE #463: Symfony 5 support
## Feature Request

Are you planning to add support for Symfon5?

### PR DESCRIPTION
**Pull request by jorrit**
## Subject

Add support for Symfony 5, Sonata Admin 4,  Twig 3.

Closes #463.

## Changelog

```markdown
### Added
- Added support for Symfony 5 / Sonata Admin 4 / Twig 3.
```
## To do

- [ ] Make sure everything works.

### CODE REVIEW

File: src/Form/Type/FormatterType.php
Comment (VincentLanglet) [inline src/Form/Type/FormatterType.php:36]: You need to update the conf isnt it ?

File: composer.json
Comment (jordisala1991) [inline composer.json]: instead of 5.1 you should use 5.3 for all symfony packages

File: composer.json
Comment (jordisala1991) [inline composer.json]: this one only 6.0

File: composer.json
Comment (jordisala1991) [inline composer.json]:
```suggestion
        "twig/twig": "^2.12.1 || ^3.0"
```

File: composer.json
Comment (jordisala1991) [inline composer.json]:
```suggestion
        "sonata-project/admin-bundle": "^4.0",
```

File: composer.json
Comment (jordisala1991) [inline composer.json:54]: you can remove conflict with core bundle

File: composer.json
Comment (jordisala1991) [inline composer.json:29]: you can leave only 1.4

**Review by jordisala1991 — CHANGES_REQUESTED** [review]

**Comment by jorrit** [comment]
Thanks for your review, @jordisala1991  !

**Comment by jordisala1991** [comment]
Can you edit the workflow file?

You need to add symfony 5.3 and 5.4. It will need also a dev-kit PR, but we would like to see if all the build pass.

It would also be nice if you try to add symfony 6.0 to the composer.json too (and on the workflow file)

**Comment by jorrit** [comment]
Is the build variant for sonata-project/block-bundle:"3.*" still relevant? If not, I think I need to create a PR on devkit.

**Comment by jorrit** [comment]
friendsofsymfony/ckeditor-bundle is not ready for Symfony 6.

**Comment by VincentLanglet** [comment]
> Is the build variant for sonata-project/block-bundle:"3.*" still relevant? If not, I think I need to create a PR on devkit.

It's not relevant, we cannot support both sonata admin 3 and 4. And Sonata admin 4 only use block 4.

File: composer.json
Comment (jordisala1991) [inline composer.json]: if ckeditor bundle is not compatible yet, please forget my previous comment about adding 6.0 to composer.json, revert to only 4.4 and 5.3

Comment (jorrit) [reply composer.json]: I reverted it.

**Comment by VincentLanglet** [comment]
There will be https://github.com/FriendsOfSymfony/FOSCKEditorBundle/pull/239 to follow

**Comment by VincentLanglet** [comment]
> There will be [FriendsOfSymfony/FOSCKEditorBundle#239](https://github.com/FriendsOfSymfony/FOSCKEditorBundle/pull/239) to follow

Merged, but their looking for a maintainer...

**Comment by jorrit** [comment]
Perhaps we should not wait for a new release for the CKEditor bundle and add support for Symfony 5 now, and Symfony 6 when CKEditor is updated.

**Comment by jordisala1991** [comment]
> Perhaps we should not wait for a new release for the CKEditor bundle and add support for Symfony 5 now, and Symfony 6 when CKEditor is updated.

Can you open a dev-kit PR with the needed changes too?

File: tests/DependencyInjection/SonataFormatterExtensionTest.php
Comment (jordisala1991) [inline tests/DependencyInjection/SonataFormatterExtensionTest.php]: Add this as a use please

Comment (jorrit) [reply tests/DependencyInjection/SonataFormatterExtensionTest.php]: ok!

**Comment by jorrit** [comment]
The dev-kit PR has been created.

File: tests/Block/FormatterBlockServiceTest.php
Comment (VincentLanglet) [inline tests/Block/FormatterBlockServiceTest.php]: This test can be removed if we remove the support of block bundle 3

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

**Comment by jorrit** [comment]
Done. I see two more tests marked `@group legacy`. When I remove those annotations, no deprecations show up in the test output. Can/should those annotations be removed in that case?

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
> Done. I see two more tests marked `@group legacy`. When I remove those annotations, no deprecations show up in the test output. Can/should those annotations be removed in that case?

Seems like there is more tests with group legacy https://github.com/sonata-project/SonataFormatterBundle/search?q=legacy ; at least on the default branch. I don't have the project locally to check if there is more.
Some of them have NEXT_MAJOR comment to help knowing what to do.

For tests without next major comment, the question would be to know why there is a group legacy. If you know why and it's not needed anymore you can remove the group. Anyway it can still be done in another PR. The stable release will be done after
- Adding param/return type everywhere
- Bumping Phpstan/psalm to good enough levels.
- Solving all the next major comment
- Mark as final as possible classes

**Comment by jorrit** [comment]
I'm working on the 5.x branch and all the NEXT_MAJOR comments have been addressed. This leaves 2 legacy groups:

 * FormatterValidatorTest::testInvalidCase: Can't find why this one is legacy.
 * SonataFormatterExtensionTest::testLoadWithoutDefaultFormatter: I don't think this test case makes sense anymore because the deprecation it tested was marked as `NEXT_MAJOR: remove this if block` and removed in 4.x.

**Review by jordisala1991 — APPROVED** [review]

**Comment by VincentLanglet** [comment]
> I'm working on the 5.x branch and all the NEXT_MAJOR comments have been addressed. This leaves 2 legacy groups:
> 
> * FormatterValidatorTest::testInvalidCase: Can't find why this one is legacy.
> * SonataFormatterExtensionTest::testLoadWithoutDefaultFormatter: I don't think this test case makes sense anymore because the deprecation it tested was marked as `NEXT_MAJOR: remove this if block` and removed in 4.x.

I've merge the dev-kit PR and this one but a new PR to remove the legacy group is welcomed.

Also, if you wanna work on the stable release of 5.0, it's welcomed too. We (@jordisala1991 and I) can help you about the TODO-list.

**Pull request merged**
